AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,236 @@
+#import <AsyncDisplayKit/ASThread.h>
+
+@interface ASMapNode()
+{
+  ASDN::RecursiveMutex _propertyLock;
+  CGSize _nodeSize;
+  MKMapSnapshotter *_snapshotter;
+  MKMapSnapshotOptions *_options;
+  CGSize _maxSize;
+  NSArray *_annotations;
+}
+@end
+
+@implementation ASMapNode
+
+@synthesize liveMap = _liveMap;
+@synthesize mapSize = _mapSize;
+@synthesize automaticallyReloadsMapImageOnOrientationChange = _automaticallyReloadsMapImageOnOrientationChange;
+@synthesize mapDelegate = _mapDelegate;
+
+- (instancetype)initWithCoordinate:(CLLocationCoordinate2D)coordinate
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+  self.backgroundColor = ASDisplayNodeDefaultPlaceholderColor();
+  _automaticallyReloadsMapImageOnOrientationChange = YES;
+  _liveMap = NO;
+
+  _options = [[MKMapSnapshotOptions alloc] init];
+  _options.region = MKCoordinateRegionMakeWithDistance(coordinate, 1000, 1000);;
+  _mapImage = [[ASImageNode alloc]init];
+  _mapImage.clipsToBounds = YES;
+  [self addSubnode:_mapImage];
+  _maxSize = self.bounds.size;
+  return self;
+}
+
+- (void)addAnnotations:(NSArray *)annotations
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (annotations.count == 0) {
+    return;
+  }
+  _annotations = [annotations copy];
+  if (annotations.count != _annotations.count && _mapImage.image) {
+    // Redraw
+    [self setNeedsDisplay];
+  }
+}
+
+- (void)setUpSnapshotter
+{
+  if (!_snapshotter) {
+    _options.size = _nodeSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+- (BOOL)isLiveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _liveMap;
+}
+
+- (void)setLiveMap:(BOOL)liveMap
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (liveMap == _liveMap) return;
+  _liveMap = liveMap;
+  liveMap ? [self addLiveMap] : [self removeLiveMap];
+}
+
+- (CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _mapSize;
+}
+
+- (void)setMapSize:(CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (CGSizeEqualToSize(mapSize,_mapSize)) {
+    return;
+  }
+  _mapSize = mapSize;
+  _nodeSize = _mapSize;
+  _automaticallyReloadsMapImageOnOrientationChange = NO;
+  [self setNeedsLayout];
+}
+
+- (BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _automaticallyReloadsMapImageOnOrientationChange;
+}
+
+- (void)setAutomaticallyReloadsMapImageOnOrientationChange:(BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (_automaticallyReloadsMapImageOnOrientationChange == automaticallyReloadsMapImageOnOrientationChange) {
+    return;
+  }
+  _automaticallyReloadsMapImageOnOrientationChange = automaticallyReloadsMapImageOnOrientationChange;
+  
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+  [self setUpSnapshotter];
+  [self takeSnapshot];
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  if (_mapView) {
+    [_mapView removeFromSupernode];
```

Comment:
It's a little unusual to remove the live map view in -clearFetchedData, but I think it may behave well in practice.  However, what happens when the map scrolls back into view?  Won't _liveMap still be set to YES, but when fetchData is called we will simply create a map snapshot image and not restore the interactive map?  I think that behavior could be OK (reverting to the non-interactive map after it is scrolled away), but we'd have to set _liveMap = NO...